### DISCUSSION TITLE
[Question] Is the Hybrid Schrödinger-Feynman backend the only one that has a multi-processing/multi-threading option?

### DISCUSSION THREAD

**Discussion by rentristandelacruz** [munich-quantum-toolkit/ddsim#245] [Q&A] [answered]
### mqt.ddsim version

1.17.3

### OS

Linux Mint 21.1 

### Python version

Python 3.10

### C++ compiler

11.3.0 

### Description

When using hybrid Schrödinger-Feynman backend (`hybrid_qasm_simulator​`), there is a `nthreads` option that you can set to specify the number of threads to use. This option does not work for the circuit simulator (`qasm_simulator`) and path simulator (`path_sim_qasm_simulator`). Also, when using `ddsim_simple`, the `--nthreads arg` option is specified to be for the hybrid simulation. I just want to confirm if the other backends/simulators do not have multi-threading/multi-processing options.

[Meta Question] Would it be better to ask ddsim-related question via email or by creating a GitHub issue? I may have other ddsim-related questions but I don't know if it's too "spammy" if I create lots of GitHub issues for my questions. But the advantage of using GitHub issues for questions is that the questions (and answers) are publicly document so they might be useful to others and one can always go back to these issues and use them as references.

Thank you!

### How to Reproduce

[Question] Is the Hybrid Schrödinger-Feynman backend the only one that has a multi-processing/multi-threading option?

**Comment by burgholzer** [comment, marked as answer]
Hey 👋🏼 

Before getting to the technical part: I converted this issue to a GitHub discussion. If you have any questions regarding DDSIM, this is probably the best place to ask since, as you already mentioned, the questions and answers are public and everyone can profit from them. Feel free to open more questions here!

As for the technical part: 
Your observation is correct. Out of the Qiskit backends, only the HSF simulator can use multithreading.
All other backends run on a single thread/core. This is mainly due to the underlying decision diagram package (https://github.com/cda-tum/dd_package) not supporting multithreading at the moment.
For some context on why multithreading is hard with decision diagrams, you can have a look at https://www.cda.cit.tum.de/files/eda/2020_aspdac_concurrency_in_dd_based_quantum_circuit_simulation.pdf
The HSF simulator circumvents these limitations by "cutting" the circuit, which creates multiple independent circuits that can then trivially be simulated in parallel. The results of each execution are then accumulated in parallel to yield to final result.

Another example of this kind of parallelization is offered by our stochastic, noise-aware simulator. 
It stochastically simulates the quantum circuit according to a given error model and averages over the results. Each of the simulations is, again, independent from all others and can, thus, be simulated in parallel.
Unforntunately, that simulator is not yet available as a Qiskit backend. 
The simulator is accessible in a similar fashion to `ddsim_simple` though. See https://github.com/cda-tum/ddsim/blob/main/apps/noise_aware.cpp

It is still active research to figure out the best way to enable multithreading for the other simulation techniques.
One really technical part of this is figuring out how multiple instances of a DD package can properly interact with each other without causing issues.

**Comment by rentristandelacruz** [comment]
Thanks for the quick reply. I'll check the paper "_Concurrency in DD-based Quantum Circuit Simulation_". For future questions, I'll use the Discussion tab instead of creating a new issue.
